### PR TITLE
Add select2 for enhanced/searchable dropdowns (1.2.x)

### PR DESCRIPTION
**Pull request by TheWitness**
## What this does

Adds **Select2** (4.1.0, MIT license, https://select2.org) as a new front-end library for enhanced/searchable `<select>` dropdowns, backported to `1.2.x` following this branch's own (manual/non-npm) vendoring conventions.

Since `1.2.x` has no `package.json`/JS build pipeline (unlike `develop`), the library is vendored directly:

- `include/js/select2.js`: the upstream `select2.full.js` build - **unminified**, matching this branch's existing convention of vendoring unminified copies (e.g. `jquery.js`, `jstree.js`). The "full" build bundles its `jquery-mousewheel` dependency, so no extra file is needed.
- `include/themes/<theme>/select2.css`: the upstream `select2.css` (structural/positioning rules), copied unmodified into all 7 themes (classic, dark, midwinter, modern, paper-plane, paw, sunrise).
- `lib/html.php`: wired both into the page header, next to the other jQuery UI plugin includes (`select2.css` after `jquery.colorpicker.css`, `select2.js` after `jquery.colorpicker.js`), matching this file's existing include style.
- `include/themes/<theme>/main.css`: Select2 4.x publishes no jQuery UI theme/adapter, so a small per-theme override block (widget border/background/text, hover, and active/highlighted colors) was appended, reusing each theme's own `jquery-ui.css`/`.ui-state-active` colors so Select2 dropdowns match that theme's native look instead of Select2's generic default styling.

## Testing

- `php -l` on all changed PHP files.
- Verified the vendored `select2.js`/`select2.css` are genuinely unminified copies of the real `select2@4.1.0` npm package (via unpkg), not `.min.js`/`.min.css`.
- Verified brace-balance across all 7 modified `main.css` files after appending the adapter block.
- Companion PR for `develop`: #8051 (uses the npm-synced pipeline that branch already has).

### CODE REVIEW

File: include/themes/classic/main.css [hunk 1]
```diff
@@ -2131,3 +2131,66 @@ td.center > table {
...
+
+.select2-container--default .select2-results__option--selected {
+	background: #dadada;
+	color: #212121;
+}
+
+.select2-container--default .select2-results__option--highlighted.select2-results__option--selectable {
+	background: #ffffff;
+	color: #212121;
+}
```
[the remark this answers is not in the document]
Comment (TheWitness) [reply include/themes/classic/main.css]: Fixed in `0cfa2ea`: changed the highlighted-option background to this theme's own hover color (`#dadada` / `#212121`) instead of white, so keyboard/pointer focus is visible again.

File: include/themes/midwinter/main.css [hunk 2]
```diff
@@ -47,3 +47,66 @@
...
+/* Select2 - adapt to look like a native Cacti/jQuery UI control.
+   Select2 4.x ships no jQuery UI theme, so these overrides reuse this
+   theme's own jquery-ui.css/main.css colors (see .ui-widget-content /
+   .ui-state-hover / .ui-state-active) to match its look and feel. */
+.select2-container--default .select2-selection--single,
+.select2-container--default .select2-selection--multiple {
+	border: 1px solid #aaaaaa;
+	border-radius: 4px;
+	background: #ffffff;
+	color: #222222;
```
[the remark this answers is not in the document]
Comment (TheWitness) [reply include/themes/midwinter/main.css]: Fixed in `0cfa2ea`: rewrote this theme's whole Select2 adapter block to use the same CSS custom properties `jquery-ui.midwinter.css` uses for its widget surface and menu focus/active state (`--border-color-superlowlight`, `--background-content`, `--text-color-lighter`, `--border-color`, `--background-lowlight`) instead of hard-coded light colors, so it now follows the theme's dark palette and its `data-theme-color` switch. Also removed a leftover duplicate copy of the block that had been introduced by mistake.

File: include/themes/paper-plane/main.css [hunk 3]
```diff
@@ -2841,3 +2841,66 @@ circle {
...
+/* Select2 - adapt to look like a native Cacti/jQuery UI control.
+   Select2 4.x ships no jQuery UI theme, so these overrides reuse this
+   theme's own jquery-ui.css/main.css colors (see .ui-widget-content /
+   .ui-state-hover / .ui-state-active) to match its look and feel. */
+.select2-container--default .select2-selection--single,
+.select2-container--default .select2-selection--multiple {
+	border: 1px solid #aaaaaa;
+	border-radius: 4px;
+	background: #ffffff;
+	color: #222222;
```
[the remark this answers is not in the document]
Comment (TheWitness) [reply include/themes/paper-plane/main.css]: Fixed in `0cfa2ea`: rewrote this theme's Select2 adapter to reuse its actual dark control colors (`#111` surface, `#d3d3d3` border, `#555555` text, matching `.ui-state-default`) instead of the generic light jQuery UI palette.

File: include/themes/paper-plane/main.css [hunk 4]
```diff
@@ -2841,3 +2841,66 @@ circle {
...
+
+.select2-container--default .select2-results__option--selected {
+	background: #dadada;
+	color: #212121;
+}
+
+.select2-container--default .select2-results__option--highlighted.select2-results__option--selectable {
+	background: #ffffff;
+	color: #212121;
+}
```
[the remark this answers is not in the document]
Comment (TheWitness) [reply include/themes/paper-plane/main.css]: Fixed in `0cfa2ea`: the highlighted/selected result now reuses this theme's existing active-state gradient (`linear-gradient(to bottom, #3a3b3c 0%, #100000 100%)`, white text, matching `.ui-state-active`) instead of a flat white background.

File: include/themes/sunrise/main.css [hunk 5]
```diff
@@ -3041,3 +3041,66 @@ circle {
...
+/* Select2 - adapt to look like a native Cacti/jQuery UI control.
+   Select2 4.x ships no jQuery UI theme, so these overrides reuse this
+   theme's own jquery-ui.css/main.css colors (see .ui-widget-content /
+   .ui-state-hover / .ui-state-active) to match its look and feel. */
+.select2-container--default .select2-selection--single,
+.select2-container--default .select2-selection--multiple {
+	border: 1px solid #aaaaaa;
+	border-radius: 4px;
+	background: #ffffff;
+	color: #222222;
```
[the remark this answers is not in the document]
Comment (TheWitness) [reply include/themes/sunrise/main.css]: Fixed in `0cfa2ea`: rewrote this theme's Select2 adapter to reuse its actual dark/transparent control palette (`#000000` surface, `rgba(255,255,255,0.2)` border, `#969BA3` text, `#93CEFF` accent, matching `.ui-widget-content`/`.ui-state-hover`), and used the theme's established green focus color (`#3C8A3C`, already used by `.ui-multiselect-checkboxes .ui-state-hover`) for the selected/highlighted results — this covers both occurrences in the file.

File: include/layout.js [hunk 6]
```diff
@@ -941,6 +941,93 @@ function applySkin() {
 	displayMessages();
 
 	renderLanguages();
+
+	$('.select2').each(function() {
```
[the remark this answers is not in the document]
Comment (TheWitness) [reply include/layout.js]: Fixed in `9b4ce43`: scoped all six select2 marker-class selectors to `select.<class>` (e.g. `select.select2`) so they only ever match the original `<select>` elements, never Select2's own generated `.select2 .select2-container` wrapper.

File: include/layout.js [hunk 7]
```diff
@@ -941,6 +941,93 @@ function applySkin() {
...
+		}
+	});
+
+	$('.select2-callback').each(function() {
+		$(this).select2({
+			ajax: {
+				type: 'post',
+				dataType: 'json',
+				delay: 250,
+				cache: false,
```
[the remark this answers is not in the document]
Comment (TheWitness) [reply include/layout.js:1032]: Fixed in `9b4ce43`: added a `processResults` mapping (`{results: data.map(item => ({id: item.id, text: item.label}))}`) so the plain `{label, value, id}` array these callback endpoints return is normalized to Select2's expected format.

File: include/layout.js [hunk 8]
```diff
@@ -941,6 +941,93 @@ function applySkin() {
...
+		} else {
+			$(this).select2({
+				tags: true
+			});
+		}
+	});
+
+	$('.select2-callback').each(function() {
+		$(this).select2({
+			ajax: {
```
[the remark this answers is not in the document]
Comment (TheWitness) [reply include/layout.js]: This was already fixed in `bfce7e4` (pushed just before this review ran, so the review was against the prior commit): `.select2-callback` sets `dropdownParent` to the `.ui-dialog` ancestor when present. In `9b4ce43` I made it unconditional, explicitly falling back to `document.body` otherwise instead of leaving it unset.

File: include/layout.js [hunk 9]
```diff
@@ -941,6 +941,100 @@ function applySkin() {
...
+		}
+	});
+
+	$('.select2-callback').each(function() {
+		var callbackUrl = $(this).data('callback');
+		var options = {
+			ajax: {
+				type: 'post',
+				dataType: 'json',
+				delay: 250,
```
[the remark this answers is not in the document]
Comment (TheWitness) [reply include/layout.js:1031]: Fixed in 973504e: added an `ajax.data` callback that sends `__csrf_magic: csrfMagicToken`, matching every other POST in this file. The nonce configured alongside this initializer is the CSP nonce, not the CSRF token, so it needed to be sent separately.

File: include/layout.js [hunk 10]
```diff
@@ -941,6 +941,104 @@ function applySkin() {
...
+					if (params.term !== undefined && params.term != '') {
+						return callbackUrl + '&search=' + encodeURIComponent(params.term) + '&page=' + (encodeURIComponent(params.page || 1));
+					} else {
+						return callbackUrl + '&page=' + (encodeURIComponent(params.page || 1));
+					}
+				},
+				processResults: function(data) {
+					return {
+						results: $.map(data, function(item) {
+							return { id: item.id, text: item.label };
```
[the remark this answers is not in the document]
Comment (TheWitness) [reply include/layout.js]: Fixed in 973504e: added a small `decodeHtmlEntities()` helper (next to `escapeString()`) and used it when mapping `item.label` to `text`, so the already-HTML-escaped label (from `html_escape()` in the callback producers) isn't escaped a second time by Select2's own `escapeMarkup`.

**Review by browniebraun — APPROVED** [review]
